pulsar-transaction/buffer/src/main/java/org/apache/pulsar/transaction/buffer/impl/TransactionCursorImpl.java

```diff
@@ -123,4 +204,377 @@ private void addTxnToCommittedIndex(TxnID txnID, long committedAtLedgerId) {
+                if (throwable != null) {
+                    recoverFuture.completeExceptionally(throwable);
+                } else {
+                    if (snapshot.status == null) {
+                        recoverFuture.complete(null);
+                    } else {
+                        switch (snapshot.status) {
+                            case START:
+                                recoverFromStart(currentPosition).whenComplete((ignore, error) -> {
+                                    checkComplete(error, recoverFuture);
+                                });
+                                break;
+                            case MIDDLE:
+                                recoverFromMiddle(snapshot).whenComplete((ignore, error) -> {
+                                    checkComplete(error, recoverFuture);
+                                });
+                                break;
+                            case END:
+                                recoverFromEnd(snapshot).whenComplete((ignore, error) -> {
+                                    checkComplete(error, recoverFuture);
+                                });
+                        }
+                    }
+                }
+            });
+
+        return recoverFuture;
+    }
+
+    private void checkComplete(Throwable error, CompletableFuture<Void> future) {
+        if (error != null) {
+            future.completeExceptionally(error);
+        } else {
+            future.complete(null);
+        }
+    }
+
+    @Getter
+    final static class PersistentTxnIndexSnapshot {
+        enum SnapshotStatus  {
+            START,
+            MIDDLE,
+            END,
+        }
+
+        SnapshotStatus status;
+        // If the status is START, the position is the position which is the transaction log doing snapshot.
+        //  If the status is others, the position is the snapshot beginning position on the cursor ledger.
+        Position position;
+        TransactionMetaImpl meta;
+
+        PersistentTxnIndexSnapshot(byte[] entry) {
+            StoredTxn txn = DataFormat.parseStoredTxn(entry);
+            switch (txn.getStoredStatus()) {
+                case START:
+                    this.status = SnapshotStatus.START;
+                    break;
+                case MIDDLE:
+                    this.status = SnapshotStatus.MIDDLE;
+                    break;
+                case END:
+                    this.status = SnapshotStatus.END;
+                    break;
+            }
+            this.meta = DataFormat.recoverMeta(txn.getTxnMeta());
+            this.position = DataFormat.recoverPosition(txn.getPosition());
+        }
+
+    }
+
+    private CompletableFuture<Void> recoverFromStart(Position currentPosition) {
+        return readPrevEntry(currentPosition)
+                   .thenApply(entry -> new PersistentTxnIndexSnapshot(entry.getData()))
+                   .thenCompose(this::recoverFromEnd);
```

Comment:
The previous position of a start entry is not the end entry. Because the previous snapshot can be an incomplete snapshot, no?